File: cdm-sources/src/cdm/margin/schedule/functions/StandardizedScheduleAssetClass.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.product.qualification.functions.Qualify_AssetClass_Equity import Qualify_AssetClass_Equity
from cdm.event.common.Trade import Trade
from cdm.margin.schedule.StandardizedScheduleAssetClassEnum import StandardizedScheduleAssetClassEnum
from cdm.product.qualification.functions.Qualify_AssetClass_InterestRate import Qualify_AssetClass_InterestRate
from cdm.product.qualification.functions.Qualify_AssetClass_Credit import Qualify_AssetClass_Credit
from cdm.product.qualification.functions.Qualify_AssetClass_Commodity import Qualify_AssetClass_Commodity
from cdm.product.qualification.functions.Qualify_AssetClass_ForeignExchange import Qualify_AssetClass_ForeignExchange

__all__ = ['StandardizedScheduleAssetClass']


@replaceable
def StandardizedScheduleAssetClass(trade: Trade) -> StandardizedScheduleAssetClassEnum:
    """
    Identifies the asset class of a trade from qualifying functions, according to the standardized schedule classification.
    
    Parameters 
    ----------
    trade : Trade
    
    Returns
    -------
    assetClass : StandardizedScheduleAssetClassEnum
    
    """
    self = inspect.currentframe()
    
    
    def _then_fn4():
        return rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "COMMODITY")
    
    def _else_fn4():
        return True
    
    def _then_fn3():
        return rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "EQUITY")
    
    def _else_fn3():
        return if_cond_fn(Qualify_AssetClass_Commodity(rosetta_resolve_attr(self, "economicTerms")), _then_fn4, _else_fn4)
    
    def _then_fn2():
        return rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "FOREIGN_EXCHANGE")
    
    def _else_fn2():
        return if_cond_fn(Qualify_AssetClass_Equity(rosetta_resolve_attr(self, "economicTerms")), _then_fn3, _else_fn3)
    
    def _then_fn1():
        return rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "CREDIT")
    
    def _else_fn1():
        return if_cond_fn(Qualify_AssetClass_ForeignExchange(rosetta_resolve_attr(self, "economicTerms")), _then_fn2, _else_fn2)
    
    def _then_fn0():
        return rosetta_resolve_attr(StandardizedScheduleAssetClassEnum, "INTEREST_RATES")
    
    def _else_fn0():
        return if_cond_fn(Qualify_AssetClass_Credit(rosetta_resolve_attr(self, "economicTerms")), _then_fn1, _else_fn1)
    
    product = rosetta_resolve_attr(rosetta_resolve_attr(self, "trade"), "product")
    economicTerms = rosetta_resolve_attr(rosetta_resolve_attr(self, "product"), "economicTerms")
    assetClass =  if_cond_fn(Qualify_AssetClass_InterestRate(rosetta_resolve_attr(self, "economicTerms")), _then_fn0, _else_fn0)
    
    
    return assetClass

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/margin/schedule/functions/StandardizedScheduleAssetClass.java
package cdm.margin.schedule.functions;

import cdm.event.common.Trade;
import cdm.margin.schedule.StandardizedScheduleAssetClassEnum;
import cdm.product.qualification.functions.Qualify_AssetClass_Commodity;
import cdm.product.qualification.functions.Qualify_AssetClass_Credit;
import cdm.product.qualification.functions.Qualify_AssetClass_Equity;
import cdm.product.qualification.functions.Qualify_AssetClass_ForeignExchange;
import cdm.product.qualification.functions.Qualify_AssetClass_InterestRate;
import cdm.product.template.EconomicTerms;
import cdm.product.template.NonTransferableProduct;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperS;
import javax.inject.Inject;


@ImplementedBy(StandardizedScheduleAssetClass.StandardizedScheduleAssetClassDefault.class)
public abstract class StandardizedScheduleAssetClass implements RosettaFunction {
	
	// RosettaFunction dependencies
	//
	@Inject protected Qualify_AssetClass_Commodity qualify_AssetClass_Commodity;
	@Inject protected Qualify_AssetClass_Credit qualify_AssetClass_Credit;
	@Inject protected Qualify_AssetClass_Equity qualify_AssetClass_Equity;
	@Inject protected Qualify_AssetClass_ForeignExchange qualify_AssetClass_ForeignExchange;
	@Inject protected Qualify_AssetClass_InterestRate qualify_AssetClass_InterestRate;

	/**
	* @param trade 
	* @return assetClass 
	*/
	public StandardizedScheduleAssetClassEnum evaluate(Trade trade) {
		StandardizedScheduleAssetClassEnum assetClass = doEvaluate(trade);
		
		return assetClass;
	}

	protected abstract StandardizedScheduleAssetClassEnum doEvaluate(Trade trade);

	protected abstract MapperS<? extends NonTransferableProduct> product(Trade trade);

	protected abstract MapperS<? extends EconomicTerms> economicTerms(Trade trade);

	public static class StandardizedScheduleAssetClassDefault extends StandardizedScheduleAssetClass {
		@Override
		protected StandardizedScheduleAssetClassEnum doEvaluate(Trade trade) {
			StandardizedScheduleAssetClassEnum assetClass = null;
			return assignOutput(assetClass, trade);
		}
		
		protected StandardizedScheduleAssetClassEnum assignOutput(StandardizedScheduleAssetClassEnum assetClass, Trade trade) {
			final Boolean boolean0 = qualify_AssetClass_InterestRate.evaluate(economicTerms(trade).get());
			if ((boolean0 == null ? false : boolean0)) {
				assetClass = StandardizedScheduleAssetClassEnum.INTEREST_RATES;
			} else {
				final Boolean boolean1 = qualify_AssetClass_Credit.evaluate(economicTerms(trade).get());
				if ((boolean1 == null ? false : boolean1)) {
					assetClass = StandardizedScheduleAssetClassEnum.CREDIT;
				} else {
					final Boolean boolean2 = qualify_AssetClass_ForeignExchange.evaluate(economicTerms(trade).get());
					if ((boolean2 == null ? false : boolean2)) {
						assetClass = StandardizedScheduleAssetClassEnum.FOREIGN_EXCHANGE;
					} else {
						final Boolean boolean3 = qualify_AssetClass_Equity.evaluate(economicTerms(trade).get());
						if ((boolean3 == null ? false : boolean3)) {
							assetClass = StandardizedScheduleAssetClassEnum.EQUITY;
						} else {
							final Boolean boolean4 = qualify_AssetClass_Commodity.evaluate(economicTerms(trade).get());
							if ((boolean4 == null ? false : boolean4)) {
								assetClass = StandardizedScheduleAssetClassEnum.COMMODITY;
							} else {
								assetClass = null;
							}
						}
					}
				}
			}
			
			return assetClass;
		}
		
		@Override
		protected MapperS<? extends NonTransferableProduct> product(Trade trade) {
			return MapperS.of(trade).<NonTransferableProduct>map("getProduct", _trade -> _trade.getProduct());
		}
		
		@Override
		protected MapperS<? extends EconomicTerms> economicTerms(Trade trade) {
			return product(trade).<EconomicTerms>map("getEconomicTerms", nonTransferableProduct -> nonTransferableProduct.getEconomicTerms());
		}
	}
}
